be/Service/StatisticsChartDataSvc.php

```diff
@@ -0,0 +1,141 @@
+<?php declare(strict_types=1);
+
+namespace Service;
+
+use Model\Entity\Game;
+use Model\Entity\User;
+use Model\Repository\GameRepository;
+use Model\Repository\UserRepository;
+
+
+class StatisticsChartDataSvc
+{
+    private const DEFAULT_USER_AMOUNT = 5;
+
+    /** @var int[][] */
+    private $data = [];
+
+    /** @var User[] */
+    private $users = [];
+
+    /** @var UserRepository */
+    private $userRepository;
+
+    /** @var GameRepository */
+    private $gameRepository;
+
+    public function __construct(
+        UserRepository $userRepository,
+        GameRepository $gameRepository
+    ) {
+        $this->userRepository = $userRepository;
+        $this->gameRepository = $gameRepository;
+    }
+
+    /**
+     * @param int[] $userNidList
+     * @return int[][]
+     */
+    public function loadData(array $userNidList): array
```

Comment:
W tej metodzie bazujesz na zmiennych globalnych klasy. Nie czytelniej i mniej bugogennie byłoby gdy by te metody: loadUsers i loadDataFromDb poprostu zwracały dane i przekazywały do następnej funkcji (taki pipeline) ?

Reply:
@kamilkuz 
Wiem. Zrobiłem tak nie bez powodu. Choć odgórnie i szczerze napiszę, że nie wiem czy pełni dobrze.

Chodzi o to, że jeśli miałbym przekazywać je poprzez argument, to musiałbym tego dokonać w dosłownie każdej metodzie tej klasy. Spowodowałoby to, że np. `setPoints()` wymagałaby aż 5 parametrów. Wujek Bob, jak zapewne wiesz, o tym co nieco wspominał. W jego książce był przykład, że jeśli dochodzi do takich sytuacji, to lepiej jest odseparować to do takiego stanu klasy. Tak mi się przynajmniej wydaje, można by ewentualnie poszukać potwierdzenia u jakiegoś źródła.
Tutaj sprawę tą jeszcze nieco ułatwia to, że te dane wykorzystywane są w każdej metodzie, i logika całej klasy nie odbiega o nic innego. Dodatkowo, jak skorzysta się z tej metody publicznej ponownie, to poprzedni stan zostanie nadpisany.

A i co ważne. Nie nazywaj "zmiennymi globalnymi klasy" tych zmiennych, bo słowo "globalne" bardzo źle się kojarzy w programowaniu :P. To są "zmienne instancyjne" albo "właściwości klasy" ;).

Reply:
@ktrzos 
No spoko, wiem że zmienne instancyjne są standardową techniką obiektową. Nazwałem je globalne bo z pewnego punktu widzenia jeśli jakaś metoda bazuje na zmiennych których nie dostaje w parametrze to ta zmienna jest w jakimś sensie globalna :P. Ale tak wiem, w obiektowym nazewnictwie to zmienne instancyjne.

Osobiście uważam że ich korzystanie powinno być ograniczone do absolutnego minimum. One według mnie wprowadzają wiele problemów:
- mniejsza czytelność - patrząc na funkcje nie wiesz na jakich danych operuje
- bugogennoś - można niezuważyć że metoda bazuje na nich, coś w niej zmienić i może się coś rozwalić
- bugogenność 2 - jakaś metoda bazuje na takich danych i wymaga że inna metoda będzie te dane uzupełniać. Wystarczy że np. zmienimy że ta metoda uzupełniająca dane zostanie wywołana w innym momencie i już mamy kolejny problem

Jeśli chodzi o ilość parametrów - wiem że duża ilość jest zła, ale to ma tą przewagę, że jest dużo przejrzystsze pod tym wzlgędem że wiesz na jakich danych dana metoda operuje.
Tak naprawdę wszystko to mówię z punktu widzenie programowania funkcyjnego, wiem że obiektowe rządzi się innymi prawami i też kojarzę że wujek Bob tak mówił więc spoko i się nie upieram.

Reply:
Tu nie kwestia upierania się :). Podczas takich rozmów i tego typu sytuacji osobiście liczę na takie konwersacje, z których zarówno Ty, jak i ja coś z tego wyniesiemy :). 

W tym przypadku nie jestem w 100% przekonany do tego rozwiązania. Tym bardziej, że sam nie przepadam za jakimikolwiek formami "globalnego" stanu.  Fakt faktem, patrzysz mocno od strony funkcyjnej :). Ja też się tak staram podchodzić do tematu, bo jednak ten sposób myślenia i mi się spodobał. Jednakże... jestem zdecydowanie przeciwny funkcjom/metodom z większą ilością parametrów. To w wielu przypadkach bardziej komplikuje kod niż wspomaga. Dlatego w tym przypadku zdecydowałem się na takie rozwiązanie.

Nie mniej jednak, przysiądę do tego jeszcze raz. Może coś uda się wymyślić, aby wyglądało to lepiej :). To w sumie jest niezła zagadka logiczna.

Reply:
@kamilkuz 
Zmodyfikowałem kod tej klasy. Będę chciał dokonać większych poprawek, aby nie występowały tam metody z większą ilością parametrów niż 2. Dziś niestety już nie dam rady tego ogarnąć.

Reply:
Nieźle, spoko

Reply:
@kamilkuz 
Jednak nie ogarnę tego w najbliższym czasie, bo ciągle mi coś wypada i nie wyrabiam. Proponuję zrobić merge (jeśli wszystko inne będzie ok), i stworzę osobnego kejsa na tą sytuację, bo pewnie będę musiał większą ilość czasu na to poświęcić.